File: erpnext_enhancements/chat/retrieval/citations.py
# Copyright (c) 2026, Sapphire Fountains and contributors
# For license information, please see license.txt

"""The citation manifest, and the URL resolution that keeps a model out of the anchor.

**Citations are assembly-order integers whose manifest is known before generation.** That is
the whole design in one sentence, and every property below follows from it: the manifest can
stream *first*, inline links resolve live as tokens arrive, and an id the model invents
resolves to nothing rather than to somebody else's room.

--------------------------------------------------------------------------------------
The model never authors a URL
--------------------------------------------------------------------------------------

The model emits ``[[ref:7]]``. It does not emit a link, and there is no code path by which
text it produced becomes an ``href``. The URL for ref 7 was built **server-side** from the
manifest before generation started.

That is not defence-in-depth, it is the actual defence. A model that can author an anchor can
author ``javascript:…``, or a link into a room the reader is not in, or a plausible-looking
link to an unrelated message — and the last of those is the dangerous one, because nothing
about it looks wrong. Resolving server-side from a manifest the gate built makes all three
unreachable rather than filtered.

**Anchors are built with DOM APIs, never ``innerHTML``.** The widget renders assistant text
through ``frappe.markdown(...)``, and whether the deployed Frappe sanitises that output is
itself an open question — so building citation anchors with ``createElement``/``textContent``
keeps them out of that blast radius entirely rather than betting on the answer.

--------------------------------------------------------------------------------------
An unknown ref is dropped silently, and counted
--------------------------------------------------------------------------------------

A model citing ``[[ref:99]]`` when the manifest has eleven entries is not a page the user
should see broken. The marker is removed and the sentence reads normally.

But it is **counted**, because a rising miss rate is a *prompt* regression signal rather than
a UI bug: the model has started inventing citations, which means it is being asked for them in
a way it cannot satisfy. Silence about that is how a slow degradation goes unnoticed for
months.

--------------------------------------------------------------------------------------
The stream tail buffer
--------------------------------------------------------------------------------------

``[[ref:11]]`` can arrive split across two SSE frames. A renderer that parses each frame in
isolation produces two broken fragments where there should be one anchor, and it does so only
sometimes — which is the worst kind of bug to reproduce. The client holds back up to
:data:`MAX_MARKER_LENGTH` characters of tail until it can prove they are not a partial marker.
That constant lives here so both sides derive it from one place.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from erpnext_enhancements.chat.links import build_chat_route

#: What the model is asked to emit, and what the client parses. Tolerant of internal spacing
#: because a model will produce ``[[ref: 3]]`` sooner or later and refusing it buys nothing.
REF_PATTERN: re.Pattern[str] = re.compile(r"\[\[\s*ref\s*:\s*(\d+)\s*\]\]")

#: The longest marker the tolerant pattern can match for a plausible id, which is how much
#: tail a streaming renderer must hold back. ``[[ref: 999 ]]`` is 14; 16 is that plus slack.
MAX_MARKER_LENGTH: int = 16

#: The label format for a context line, which is what the model sees. The timestamp and the
#: zone are both present on purpose: "yesterday at 2" is ambiguous across a team in two
#: timezones, and a model asked to reason about ordering needs the offset rather than a
#: friendly string.
CONTEXT_LINE_TEMPLATE: str = "⟦ref:{ref}⟧ {author} ({timestamp}): {body}"


#: The client's citation types, from ``public/js/chat/citations.js``. **This is a contract
#: with code that already shipped**, not a vocabulary this module gets to choose:
#: ``citations.js`` has a ``LINKABLE`` set and renders anything outside it as a flat,
#: non-navigating pill.
#:
#: ``digest`` is deliberately absent from ``LINKABLE`` on the client side, and that absence is
#: right: a digest has no single canonical message to point at, and a link that guesses one is
#: a lie. It is a weaker claim than "this message said so" and it renders as one.
TYPE_CHAT_MESSAGE: str = "chat_message"
TYPE_DIGEST: str = "digest"

#: ``kind`` (how this module thinks about a source) → ``type`` (what the client renders).
#: A chunk *is* a run of messages, so it points at the first of them; the reader lands on
#: real text they can check rather than on a synthetic object.
_KIND_TO_TYPE: dict[str, str] = {
	"chunk": TYPE_CHAT_MESSAGE,
	"message": TYPE_CHAT_MESSAGE,
	"digest": TYPE_DIGEST,
}


@dataclass(frozen=True)
class Citation:
	"""One manifest entry. ``ref`` is its position in assembly order, 1-based."""

	ref: int
	room: str
	message: str | None
	label: str
	#: ``chunk`` / ``message`` / ``digest``. Translated to the client's vocabulary by
	#: :meth:`as_dict`; nothing outside this module should read it.
	kind: str = "chunk"
	subtitle: str = ""
	thread_root: str | None = None
	url: str = ""

	def as_dict(self) -> dict[str, object]:
		"""The wire shape, **matching the renderer that already shipped in Phase 3**.

		``k`` rather than ``ref``, and ``type`` rather than ``kind``, because
		``public/js/chat/citations.js`` was written against that contract in Phase 3 and its
		``indexManifest`` silently **discards** any entry whose ``k`` is not a positive
		integer. An entry keyed on ``ref`` would therefore not fail — it would vanish, every
		citation would be a miss, and the answer would render with its markers quietly
		stripped. ``tests/test_chat_api_contracts.py`` asserts the two sides agree.

		**No ``snippet``.** The client will render one in the tooltip if present, and a
		snippet is message text — putting it here would carry conversation into
		``Triton Invocation Log``, which is otherwise content-free by construction. The label
		(who, and which room) is what a reader needs to decide whether to click, and clicking
		is what shows them the message under the full permission check.
		"""
		return {
			"k": self.ref,
			"type": _KIND_TO_TYPE.get(self.kind, TYPE_CHAT_MESSAGE),
			"label": self.label,
			"url": self.url,
			"room": self.room,
			"message": self.message,
		}


def build_manifest(entries: list[dict[str, object]]) -> list[Citation]:
	"""Number the assembled sources in **assembly order**, 1-based.

	Assembly order rather than rank order, and the distinction matters: the model sees the
	context lines in assembly order, so a manifest numbered by rank would have the model
	citing ``[[ref:1]]`` for the third thing it read. Numbering what it actually sees is what
	makes the reference resolvable at all.

	Refs are **1-based** because they appear in prose. ``[[ref:0]]`` reads as a bug to a human
	and models trained on human text produce 1-based references whatever the prompt says.
	"""
	manifest: list[Citation] = []
	for index, entry in enumerate(entries, start=1):
		room = str(entry.get("room") or "")
		message = entry.get("message")
		manifest.append(
			Citation(
				ref=index,
				room=room,
				message=str(message) if message else None,
				label=str(entry.get("label") or f"message in {room}"),
				kind=str(entry.get("kind") or "chunk"),
				subtitle=str(entry.get("subtitle") or ""),
				thread_root=str(entry.get("thread_root")) if entry.get("thread_root") else None,
				url=str(entry.get("url") or ""),
			)
		)
	return manifest


def resolve_urls(manifest: list[Citation]) -> list[Citation]:
	"""Fill each entry's ``url`` from its room and message, server-side.

	Uses the same ``links`` helper as the SPA router and the notification deep link — one
	function, three consumers, which is why it was written in Phase 1 rather than at the first
	place that needed it. A fourth route table here would diverge from the other three, and
	the symptom would be citations that 404 for reasons nobody can reproduce from the answer.

	Called from a request context because the absolute origin comes from Frappe's own site-URL
	helper rather than from a settings field — a settings copy of the site host is wrong on
	exactly the day somebody changes the host.
	"""
	from erpnext_enhancements.chat.links import build_message_deep_link

	resolved: list[Citation] = []
	for entry in manifest:
		if entry.url:
			resolved.append(entry)
			continue
		try:
			url = build_message_deep_link(entry.room, entry.message, entry.thread_root)
		except Exception:
			# A citation with no URL renders as inert text rather than a link. Losing a link
			# is a cosmetic failure; raising here would lose the whole answer.
			url = ""
		resolved.append(
			Citation(
				ref=entry.ref,
				room=entry.room,
				message=entry.message,
				label=entry.label,
				kind=entry.kind,
				subtitle=entry.subtitle,
				thread_root=entry.thread_root,
				url=url,
			)
		)
	return resolved


def relative_route(room: str, message: str | None = None, thread: str | None = None) -> str:
	"""The in-app route, for a client that would rather not follow an absolute URL.

	Bench-free, unlike :func:`resolve_urls` — this is the half of the route table that needs
	no site origin, which is what lets the route shape be asserted on both sides without a
	bench.
	"""
	return build_chat_route(room, message, thread)


def context_line(citation: Citation, author: str, timestamp: str, body: str) -> str:
	"""One labelled line of context, as the model sees it.

	The marker in the *context* uses mathematical white brackets while the marker the model is
	asked to *emit* uses double square brackets. Two different shapes on purpose: it makes
	"the model echoed the label it was shown" distinguishable from "the model produced a
	citation", which is the difference between working and appearing to work.
	"""
	return CONTEXT_LINE_TEMPLATE.format(
		ref=citation.ref,
		author=author or "unknown",
		timestamp=timestamp or "unknown time",
		body=body or "",
	)


def cited_refs(text: str) -> list[int]:
	"""Every ref the model emitted, in order of first appearance, de-duplicated."""
	seen: set[int] = set()
	order: list[int] = []
	for match in REF_PATTERN.finditer(text or ""):
		ref = int(match.group(1))
		if ref in seen:
			continue
		seen.add(ref)
		order.append(ref)
	return order


def split_known_refs(text: str, manifest: list[Citation]) -> tuple[list[int], list[int]]:
	"""``(resolvable, missing)`` — the second list is the ``citation_miss_count``."""
	known = {entry.ref for entry in manifest}
	resolvable: list[int] = []
	missing: list[int] = []
	for ref in cited_refs(text):
		(resolvable if ref in known else missing).append(ref)
	return resolvable, missing


def strip_unknown_refs(text: str, manifest: list[Citation]) -> str:
	"""Remove markers that resolve to nothing, leaving the prose intact.

	Adjacent whitespace is collapsed so a removed marker does not leave a double space
	mid-sentence — the point of dropping silently is that the reader cannot tell, and a
	visible gap tells them.
	"""
	known = {entry.ref for entry in manifest}

	def _replace(match: re.Match[str]) -> str:
		return "" if int(match.group(1)) not in known else match.group(0)

	cleaned = REF_PATTERN.sub(_replace, text or "")
	return re.sub(r"[ \t]{2,}", " ", cleaned)

def flatten_refs(text: str) -> str:
	"""``[[ref:7]]`` → ``[7]``, for a surface that cannot render a chip.

	The ERPNext SPA turns each marker into a clickable chip against the manifest. Google Chat
	has neither the manifest nor a chip, so it received the raw marker and displayed
	``[[ref:25]]`` verbatim — which reads as a bug in the bot rather than as a citation.

	**Flattened rather than stripped, and that is the judgement call.** Removing them gives
	cleaner prose, but *"encountering a series of invocation errors [[ref:25]], [[ref:12]]"*
	becomes *"…errors ,"* — the sentence was written around the markers. Keeping the number
	also keeps the two surfaces consistent for anyone reading both, and the same message in
	ERPNext shows exactly ``[25]``.

	What it cannot do is make the number resolvable *in Google Chat*: there is no sources list
	there, so ``[25]`` is a pointer into something the reader has to open ERPNext to see. Making
	it a real link would mean persisting the manifest on ``Chat Message``, which is a schema
	change and a bigger decision than a rendering fix.

	Pure, and applied at the relay boundary only — the stored ``text`` keeps the markers,
	because the SPA needs them to place its chips.
	"""
	return REF_PATTERN.sub(lambda match: f"[{match.group(1)}]", text or "")

File: erpnext_enhancements/public/js/chat/citations.js
/**
 * Inline `[[ref:N]]` citation rendering. Decision #7's addition, and its whole contract.
 *
 * THE THREE RULES THIS FILE EXISTS TO KEEP
 * ========================================
 *
 * 1. **The sources dropdown is unchanged.** It is populated from the manifest, not from
 *    what the model cited, so a stream with no citations at all renders exactly as it does
 *    today and the dropdown is exactly as full as it was. Nothing in this module touches it.
 * 2. **Streaming survives.** The tokenizer is fed the newly-flushable segment only, never
 *    the whole accumulated answer — re-running a regex over the full string on every frame
 *    is O(n²) and visibly stutters on long answers.
 * 3. **No `innerHTML` string concatenation, anywhere.** Labels and snippets are
 *    user-authored: a coworker's message quoted as a source, a filename, a document title.
 *    Concatenating them into HTML is a stored-XSS vector with a straight path from any
 *    employee to every employee. Every node here is built with `document.createElement` and
 *    `textContent`.
 *
 * WHAT MAKES IT SHIPPABLE BEFORE PHASE 5
 * ======================================
 *
 * Phase 5 builds the retrieval that produces the manifest. This module is written against
 * the contract and **degrades to today's exact behaviour when no `citations` event
 * arrives** — that is the required behaviour, not a fallback afterthought, and it is what
 * lets this phase be accepted and shipped independently. `applyCitations` with an empty
 * manifest is a no-op that returns the container untouched.
 *
 * Every exported function except `applyCitations` is pure (no DOM, no clock, no network),
 * so `scripts/test_chat_citations.js` runs the whole degradation table under plain `node`.
 */

/** Tolerant match. The model may space it oddly; anything that does not match stays literal
 *  text, because the model may legitimately be *discussing* the syntax. */
export const CITATION_RE = /\[\[\s*ref\s*:\s*(\d+)\s*\]\]/g;

/**
 * The longest a citation token can be, in characters, used to size the streaming tail
 * buffer. `[[ref:` + digits + `]]` with tolerant whitespace; 16 covers every id this system
 * can assign (manifests are 1..N over a bounded context) with room to spare.
 */
export const MAX_TOKEN_LEN = 16;

/** Citation types whose `url` we will render as a link. `digest` is deliberately absent. */
const LINKABLE = new Set(["erpnext_doc", "chat_message", "email", "drive_file", "url"]);

/**
 * Turn a manifest array into a lookup keyed on `k`.
 *
 * Ids are small integers assigned at context-assembly time, `1..N`, which is what makes an
 * id outside that range PROVABLY invalid and therefore safely droppable. Entries without a
 * usable `k` are discarded here rather than at render time.
 */
export function indexManifest(entries) {
	const index = new Map();
	for (const entry of entries || []) {
		if (!entry || typeof entry !== "object") continue;
		const k = Number(entry.k);
		if (!Number.isInteger(k) || k < 1) continue;
		index.set(k, entry);
	}
	return index;
}

/**
 * **Pure.** Split a text segment into `{type:"text"}` / `{type:"citation"}` tokens.
 *
 * The unit `§4.10` asks for: "Input: a text segment plus the manifest. Output: a token list
 * that the component turns into DOM." Keeping the split and the DOM construction apart is
 * what makes the whole degradation table testable without a browser.
 *
 * An id that is not in the manifest is **dropped silently** — the token disappears and the
 * surrounding prose is left intact. Not rendered raw (which looks like the model leaked its
 * scaffolding), not rendered as a dead link (which is worse: it looks like a source until
 * clicked). `onMiss` is called so the caller can count them; a `citation_miss` rate above
 * ~2% over a rolling window means a prompt edit broke citing, silently, and is worth a
 * health check.
 */
export function tokenizeCitations(text, manifest, onMiss) {
	const source = String(text == null ? "" : text);
	const index = manifest instanceof Map ? manifest : indexManifest(manifest);
	const out = [];
	if (!source) return out;

	// NO MANIFEST AT ALL is a different thing from a manifest that lacks this id, and the two
	// must not be collapsed:
	//
	//   * manifest present, id missing  -> the model invented an id. Drop the token.
	//   * manifest absent               -> the FEATURE is off (Phase 5 is not emitting yet, or
	//                                      the flag is down). Touch nothing. The body renders
	//                                      exactly as it does today, including a stray token,
	//                                      because today nothing rewrites it either.
	//
	// The rule lives here, at the lowest level, rather than only in `applyCitations`, so no
	// caller can reach the drop path with an empty manifest by accident. This is the row that
	// lets Phase 3 ship before Phase 5.
	if (!index.size) return [{ type: "text", value: source }];

	let last = 0;
	CITATION_RE.lastIndex = 0;
	let match;
	while ((match = CITATION_RE.exec(source)) !== null) {
		const k = Number(match[1]);
		const entry = index.get(k);
		if (match.index > last) out.push({ type: "text", value: source.slice(last, match.index) });
		if (entry) {
			out.push({ type: "citation", k, citation: entry });
		} else if (typeof onMiss === "function") {
			onMiss(k);
		}
		last = match.index + match[0].length;
	}
	if (last < source.length) out.push({ type: "text", value: source.slice(last) });

	// Coalesce adjacent text runs so a dropped token does not leave the prose split into two
	// nodes with a seam a screen reader announces as two separate phrases.
	return coalesce(out);
}

/**
 * The streaming tail buffer. A citation token can arrive split across chunks — `[[re` then
 * `f:7]]` — so text is only released once it cannot be the prefix of a token.
 *
 * `flush()` on stream end is NOT optional. Forgetting it truncates the last few characters
 * of every answer that happens to end near a token, which is maddening to diagnose because
 * it depends on where the model stopped.
 */
export class CitationStream {
	constructor() {
		this.buffer = "";
	}

	/**
	 * Feed a chunk; get back the portion that is safe to render now.
	 * @returns {string} possibly empty
	 */
	push(chunk) {
		this.buffer += String(chunk == null ? "" : chunk);
		const hold = holdbackLength(this.buffer);
		if (hold === 0) {
			const out = this.buffer;
			this.buffer = "";
			return out;
		}
		const out = this.buffer.slice(0, this.buffer.length - hold);
		this.buffer = this.buffer.slice(this.buffer.length - hold);
		return out;
	}

	/** Release everything held back. Call on `done` and on `error`, always. */
	flush() {
		const out = this.buffer;
		this.buffer = "";
		return out;
	}
}

/**
 * How many trailing characters could still be the start of a citation token.
 *
 * Exported because it is the one genuinely fiddly piece of arithmetic in the file and it
 * deserves its own test rows: `"…text[["` holds 2, `"…text[[ref:7"` holds 8,
 * `"…text[[ref:7]]"` holds 0 (it is complete — the tokenizer will take it), and a lone `"["`
 * at the very end holds 1.
 */
export function holdbackLength(text) {
	const s = String(text == null ? "" : text);
	const window = s.slice(Math.max(0, s.length - MAX_TOKEN_LEN));
	for (let start = 0; start < window.length; start++) {
		const tail = window.slice(start);
		if (isTokenPrefix(tail)) return tail.length;
	}
	return 0;
}

/**
 * Whether `s` could grow into a complete `[[ref:N]]`. A COMPLETE token is **not** a prefix —
 * it is done, and holding it back would stall the stream forever on an answer that ends in a
 * citation, which is exactly the answer most likely to.
 *
 * Written as an explicit character walk rather than a nested-optional regex. The regex form
 * of "every prefix of this pattern" is unreadable, and unreadable is how the `[[ref:` case
 * ends up excluded and every stream that pauses mid-token loses six characters.
 */
export function isTokenPrefix(s) {
	const n = s.length;
	let i = 0;
	const ws = () => {
		while (i < n && /\s/.test(s[i])) i++;
	};

	if (n === 0 || s[i] !== "[") return false;
	i++;
	if (i === n) return true; // "["
	if (s[i] !== "[") return false;
	i++;
	if (i === n) return true; // "[["

	ws();
	if (i === n) return true; // "[[  "

	for (const ch of "ref") {
		if (i === n) return true; // "[[r", "[[re"
		if (s[i] !== ch) return false;
		i++;
	}
	if (i === n) return true; // "[[ref"

	ws();
	if (i === n) return true;
	if (s[i] !== ":") return false;
	i++;
	if (i === n) return true; // "[[ref:"

	ws();
	if (i === n) return true;

	let digits = 0;
	while (i < n && s[i] >= "0" && s[i] <= "9") {
		i++;
		digits++;
	}
	if (i === n) return digits > 0; // "[[ref:7"
	// "[[ref:]]" has no digits and can never become valid — it is malformed, and malformed
	// tokens stay literal text rather than being held back.
	if (digits === 0) return false;

	ws();
	if (i === n) return true;
	if (s[i] !== "]") return false;
	i++;
	if (i === n) return true; // "[[ref:7]"
	if (s[i] !== "]") return false;
	// The second "]" completes it. Complete is not a prefix.
	return false;
}

/**
 * Render a token list into a parent node. **The only impure function here.**
 *
 * Anchors are built with `createElement` and `textContent`. `href` comes from the server's
 * resolver (`citation.url`) — the client never constructs a URL per type, because desk
 * routing and doctype slugification have changed between versions and a client that builds
 * `/app/{slug}/{name}` itself breaks on an upgrade nobody connected to chat.
 *
 * `digest` renders as a NON-NAVIGATING pill. A digest has no single canonical message to
 * point at, and a link that guesses one is a lie.
 *
 * A `javascript:` (or `data:`, or anything not http/https and not same-origin-relative) URL
 * is refused and the citation degrades to a plain pill. The server validates the scheme too;
 * this is the second of two, because the first one is the one that gets refactored.
 */
export function renderTokens(parent, tokens, options) {
	const opts = options || {};
	const doc = parent.ownerDocument || document;
	for (const token of tokens) {
		if (token.type === "text") {
			parent.appendChild(doc.createTextNode(token.value));
			continue;
		}
		// `onCite` is what lets the sources row mark which entries the model ACTUALLY used
		// (approved 2026-08-10; see `orderManifestForDisplay`). It fires per rendered token,
		// so it fires repeatedly across streaming frames for the same id — the caller
		// accumulates into a Set, which is the only shape that is correct under a renderer
		// that rebuilds the bubble from scratch on every frame.
		if (typeof opts.onCite === "function") opts.onCite(token.k, token.citation);
		parent.appendChild(buildCitationNode(doc, token, opts));
	}
	return parent;
}

/**
 * **Pure.** Order a manifest for the sources row: **cited entries first, then the rest**,
 * each group in id order, with a `cited` flag on every entry.
 *
 * WHY THIS IS A BEHAVIOUR CHANGE AND WHY IT IS DELIBERATE. Locked decision #7 says the
 * sources dropdown is "preserved exactly", and research 03 §12.6 flagged the counter-proposal
 * — show the full manifest with cited entries marked and sorted first — as needing an
 * explicit human decision rather than a unilateral one. It was raised at the Phase 3
 * checkpoint and **approved on 2026-08-10**. This function is where that approval lives.
 *
 * Two properties worth stating because they are what make the change safe:
 *
 * * **With no manifest there is nothing to order**, and the caller falls back to rendering
 *   the `sources` event exactly as it did before this phase. The old path is not merely
 *   reachable, it is the path taken on every site until Phase 5 emits a manifest.
 * * **Id order, not relevance order, within each group.** The ids are assigned at
 *   context-assembly time and the inline markers in the answer read `[1]`, `[2]`, `[3]`; a
 *   sources row sorted by anything else makes the reader hunt for the chip matching the
 *   number they just clicked past.
 *
 * **The `cited` flag and the SORT are separately controllable**, and that separation is
 * load-bearing rather than tidy. While an answer is still streaming the caller wants the
 * marks to be accurate — a chip lighting up as the model leans on it is the thing worth
 * watching — but it does **not** want the row re-sorted, because `cited` grows with every
 * frame and a mid-stream sort slides chips out from under the reader's cursor. So the caller
 * passes `sortCitedFirst: false` while streaming and `true` once at the end.
 *
 * @param {Map|Array} manifest
 * @param {Set|Array} cited  ids that appeared in the rendered answer
 * @param {{sortCitedFirst?: boolean}} [options] defaults to sorting cited-first
 * @returns {Array<{k, citation, cited}>}
 */
export function orderManifestForDisplay(manifest, cited, options) {
	const index = manifest instanceof Map ? manifest : indexManifest(manifest);
	const used = cited instanceof Set ? cited : new Set(cited || []);
	const sortCitedFirst = !options || options.sortCitedFirst !== false;

	const entries = Array.from(index.entries(), ([k, citation]) => ({
		k,
		citation,
		cited: used.has(k),
	}));

	// Stable within each group: sort on (not cited, k). Comparing booleans by subtraction
	// after coercion keeps this a single comparator rather than two passes and a concat.
	if (sortCitedFirst) entries.sort((a, b) => (a.cited === b.cited ? a.k - b.k : a.cited ? -1 : 1));
	else entries.sort((a, b) => a.k - b.k);
	return entries;
}

/**
 * Rewrite the text nodes of an already-rendered body, replacing citation tokens with
 * anchors. Used on the markdown-rendered Triton bubble, where the HTML was produced by the
 * EXISTING renderer and its sanitiser policy — which is load-bearing and must not be
 * loosened to make inline links work.
 *
 * Walking text nodes rather than pre-processing the markdown source is what preserves that:
 * the sanitiser sees exactly the string it saw before this phase, and the citations are
 * added afterwards as DOM. Nothing about the existing policy changes.
 *
 * `code`, `pre` and heading elements are skipped — the system prompt tells the model not to
 * cite inside them, and if it does anyway the token is more likely to be the model
 * discussing the syntax than citing a source.
 */
export function applyCitations(container, manifest, options) {
	if (!container) return container;
	const index = manifest instanceof Map ? manifest : indexManifest(manifest);
	if (!index.size) return container; // No manifest: today's behaviour, byte for byte.

	const doc = container.ownerDocument || document;
	const skip = /^(CODE|PRE|H1|H2|H3|H4|H5|H6|A|SCRIPT|STYLE)$/;
	const walker = doc.createTreeWalker(container, 4 /* NodeFilter.SHOW_TEXT */, {
		acceptNode(node) {
			for (let el = node.parentNode; el && el !== container; el = el.parentNode) {
				if (el.nodeType === 1 && skip.test(el.nodeName)) return 2 /* FILTER_REJECT */;
			}
			return node.nodeValue && node.nodeValue.indexOf("[[") >= 0 ? 1 : 2;
		},
	});

	const targets = [];
	let node;
	while ((node = walker.nextNode())) targets.push(node);

	for (const text of targets) {
		const tokens = tokenizeCitations(text.nodeValue, index, options && options.onMiss);
		if (tokens.length === 1 && tokens[0].type === "text") continue;
		const fragment = doc.createDocumentFragment();
		renderTokens(fragment, tokens, options);
		text.parentNode.replaceChild(fragment, text);
	}

	return container;
}

/** The label a citation renders as. Falls back down the shape rather than showing "[object]". */
export function citationLabel(citation) {
	const c = citation || {};
	return String(c.label || c.name || c.title || c.url || "source");
}

/** The only two schemes that may reach an `href`. Everything else renders as inert text. */
const SAFE_PROTOCOLS = new Set(["http:", "https:"]);

/** Does the value carry its own scheme? Matches the URL spec's scheme grammar, not a guess. */
const HAS_SCHEME_RE = /^[a-zA-Z][a-zA-Z0-9+.\-]*:/;

/**
 * The origin a relative value is resolved against. Deliberately a `.invalid` host rather than
 * the real site: this is a comparison base, not a place, and hardcoding it keeps the function
 * pure — `scripts/test_chat_citations.mjs` runs it under plain `node` with no `location`.
 */
const RESOLVE_BASE = "https://url-safety-check.invalid";

/**
 * Whether a resolved URL is safe to put in an `href`.
 *
 * **Parse, do not pattern-match.** The previous implementation tested prefixes, and a prefix
 * test cannot answer this question: `/\evil.example` starts with `/`, does not start with
 * `//`, and every browser resolves it to `http://evil.example/` — a protocol-relative escape
 * straight past the one line that existed to prevent it. Backslash is not the only such trick,
 * which is the argument for handing the string to the same parser the browser will use rather
 * than trying to enumerate them: `JaVaScRiPt:`, a leading space, an embedded tab and
 * percent-encoded control characters are all normalised by the parser before we see a protocol.
 *
 * Three rules, in order:
 *
 * 1. **Scheme allowlist.** `http:` and `https:` only. `javascript:`, `data:`, `vbscript:`,
 *    `blob:`, `file:` and anything unlisted fail closed. `mailto:` is deliberately absent —
 *    see the note on `email` citations below.
 * 2. **No credentials.** `https://evil.example@sapphirefountains.com/` reads as the second
 *    host to a person and resolves to the first in some parsers. Nothing legitimate cites a
 *    URL carrying userinfo, so the ambiguity is refused rather than rendered.
 * 3. **A relative value must stay on our origin.** Relative means "starts with `/`", as
 *    before; the parser then has to agree it did not leave. `//host`, `/\host` and `/\/host`
 *    all resolve to a different origin, which is precisely the check a prefix test cannot make.
 *
 * `mailto:` is not allowed. Phase 6 §4.G.7 item 1 asks for the decision explicitly: an `email`
 * citation's `url` is built by this app and points at the ERPNext document, so it is already a
 * site-relative path; a user-typed `mailto:` gains nothing and widens the allowlist for it.
 */
export function isSafeUrl(url) {
	const value = String(url == null ? "" : url).trim();
	if (!value) return false;

	let parsed;
	try {
		parsed = new URL(value, RESOLVE_BASE);
	} catch {
		// An unparseable value is not a link. `URL` throws only on genuinely malformed input.
		return false;
	}

	if (!SAFE_PROTOCOLS.has(parsed.protocol)) return false;
	if (parsed.username || parsed.password) return false;
	if (HAS_SCHEME_RE.test(value)) return true;
	return value.startsWith("/") && parsed.origin === RESOLVE_BASE;
}

// --------------------------------------------------------------------------- internals

function buildCitationNode(doc, token, opts) {
	const citation = token.citation || {};
	const type = String(citation.type || "");
	const label = citationLabel(citation);
	const linkable = LINKABLE.has(type) && isSafeUrl(citation.url);

	const node = doc.createElement(linkable ? "a" : "span");
	node.className = "ee-citation ee-citation-" + (type || "unknown") + (linkable ? "" : " is-flat");
	node.textContent = "[" + token.k + "]";
	// title, not a tooltip library: the snippet is user-authored text and `title` is the one
	// attribute that renders it without any parsing at all.
	node.setAttribute("title", citation.snippet ? `${label} — ${citation.snippet}` : label);
	node.setAttribute("aria-label", label);
	node.dataset.citationK = String(token.k);

	if (linkable) {
		node.setAttribute("href", citation.url);
		if (/^https?:\/\//i.test(String(citation.url))) {
			node.setAttribute("target", "_blank");
			node.setAttribute("rel", "noopener noreferrer");
		}
		if (typeof opts.onNavigate === "function") {
			node.addEventListener("click", (ev) => opts.onNavigate(ev, citation));
		}
	} else if (type === "digest" && typeof opts.onExpandDigest === "function") {
		node.setAttribute("role", "button");
		node.setAttribute("tabindex", "0");
		node.addEventListener("click", () => opts.onExpandDigest(citation));
		node.addEventListener("keydown", (ev) => {
			if (ev.key === "Enter" || ev.key === " ") {
				ev.preventDefault();
				opts.onExpandDigest(citation);
			}
		});
	}

	return node;
}

function coalesce(tokens) {
	const out = [];
	for (const token of tokens) {
		const previous = out[out.length - 1];
		if (token.type === "text" && previous && previous.type === "text") {
			previous.value += token.value;
		} else {
			out.push(token);
		}
	}
	return out;
}
